accepted/2021/InboxSourceGenerators.md

```diff
@@ -0,0 +1,191 @@
+
+Alternative designs are listed in the [Q & A](#Q%20&%20A) section below.
+
+### Source generator conflict resolution
+
+Source generators and analyzers are passed to the compiler as `@(Analyzer)` items.  These items may come from NuGet packages, be built into the SDK targets, be directly defined by the user via `ProjectReference` or raw `Analyzer` items, or be provided by the new framework mechanism described in this document.  All should be considered for conflict resolution based on the same constraint.
+
+The .NET SDK already does conflict resolution for other asset types: references and copy-local runtime assets.  It should do the same for source generators, or more generally Analyzers.  Analyzers need not have "platform manifest" inputs for conflict resolution since all analyzers will be available during build, unlike runtime files which may not be available.  Comparison for the sake of conflict resolution can consider the same assembly and file version rules as reference files.  When finding two or more source generators with the same name conflict resolution will select the generator with higher assembly version, then higher file version, then that which is in the framework.  This parallels the same rules followed by reference assemblies.  The SDK may choose to optimize this comparison as it does for reference assemblies, by including a data file that includes analyzer metadata.
+
+Source generator conflict resolution should run before compile, but after all NuGet package assets have been evaluated.  Specific sequencing, public/private target naming, extensibility will not be specified here but should follow best practice as determined by the SDK team. 
+
+### Source generator versioning
+
+Source generators should be strongly named and update assembly version with every public stable release (GA and servicing).  This ensures that the publicly shipped source generators can be loaded side-by-side in the same process and follows our versioning guidelines for components which ship and must run on .NET Framework.  Source generators need not update assembly version during pre-release but should update file version to facilitate well behaved distribution by installers.  In general source generators need to follow all the same guidelines as other code that ships out of dotnet/runtime and runs outside the shared framework.
+
+### Source generator compatibility concerns
+
+Source generators should only generate code which uses public API in framework components they extend.  This public API already needs to meet the strict binary compatibility guarantees as framework API.  As such libraries containing source-generated code will be binary-compatible with future releases, just as if the user wrote this code directly.
```

Comment:
>As such libraries containing source-generated code will be binary-compatible with future releases, just as if the user wrote this code directly.

This does impose a non-obvious breaking change burden in the event we make breaking changes. We break rarely but now that we are following semantic versioning here, major versions can break when appropriate. Thoughts on how to track this? Are facilities needed to fail in predicable ways to help people understand these breaking changes.

Reply:
@terrajobst @jeffhandley 
This burden is the same we already impose on framework API, there's nothing new being imposed here.  This is just reminding folks that source-generators aren't some sort of loophole that's somehow less compatible.  If you have a source generator that you imagine isn't ready to meet those requirements then it probably shouldn't be put inbox, just like we don't put framework API in the box until we're certain it can meet those compatibility requirements.

Sure we may get things wrong and need to [obsolete](https://github.com/dotnet/designs/blob/9a6977bc6909049320695932d1a74a3a13c37a35/accepted/2020/better-obsoletion/better-obsoletion.md) API used by the generator.  I'm not seeing source generators having much of a difference in how this would work: we don't dig into binaries that aren't rebuilt to find obsolete APIs though we could and it would find source generated usage same as direct usage.  We can also obsolete the "entry point" API that source uses to interact with the generator.  If we were to do such a thing we might want to decide if the source-generator immediately stopped generating code in response to the obsoleted entry-point API or we phased it out like we do with framework API.  I think that'd need to be a context sensitive discussion that depends on the cost to the developer.